Add tests for process_reporting_metadata_staging code (minor refactor)

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-15077

Alternative to https://github.com/dimagi/commcare-hq/pull/33730

After initially thinking we wanted to handle exceptions gracefully, we have now decided that that will only mask errors further, whereas if we setup proper monitors in datadog, a task failing hard will be easier to catch.

With that in mind, I took the changes from that initial PR that I think are still good (tests and minor refactors). This PR aims to remain functionally equivalent.

Review by commit 🐟

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Added automated tests.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Encapsulate main processing code in private method

gherceg committed:
Add tests for process_reporting_metadata_staging

gherceg committed:
Use retry_on decorator instead of nested exception handling

gherceg committed:
Split test that requires TransactionTestCase into new class

This reduces running all ProcessReportingMetadataStaging tests
by ~8 seconds which seems worth it to me.

millerdev commented on corehq/apps/users/tests/test_tasks.py:
```diff
@@ -176,3 +184,76 @@ def _send_case_to_es(
+@mock.patch.object(UserReportingMetadataStaging, 'process_record')
+class TestProcessReportingMetadataStagingTransaction(TransactionTestCase):
+    """
+    TransactionTestCase is much slower than TestCase, but needed to use new_db_connection
```
Why is it a requirement to use `TransactionTestCase` with `new_db_connection`? [This test](https://github.com/dimagi/commcare-hq/blob/81f3a488487ec6e2aa60c33c8c0d795c0d1ac9e8/corehq/apps/fixtures/tests/test_upload.py#L656) doesn't do that.

gherceg replied:
It is for `select_for_update`, not `new_db_connection`. From [django docs](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#select-for-update)
> Warning
>
> Although select_for_update() normally fails in autocommit mode, since [TestCase](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.TestCase) automatically wraps each test in a transaction, calling select_for_update() in a TestCase even outside an [atomic()](https://docs.djangoproject.com/en/3.2/topics/db/transactions/#django.db.transaction.atomic) block will (perhaps unexpectedly) pass without raising a TransactionManagementError. To properly test select_for_update() you should use [TransactionTestCase](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.TransactionTestCase).

millerdev replied:
I see. The docstring says "needed to use new_db_connection"

However, after reading those docs I still have my doubts that `TransactionTestCase` is required. I think you could do the `select_for_update` in the `new_db_connection` block, and all should be fine since that connection will be closed (and its transaction ended, which terminates `select_for_update`) when the block exits.

gherceg replied:
Oh my bad! Thanks for catching that. Yeah I agree with you although I think the `select_for_update` needs to be outside of the `new_db_connection`. The purpose of the `new_db_connection` is to ensure the method under test uses a different db connection from the one that holds the lock on a record when attempting to fetch records from the db. But as far as TestCase vs TransactionTestCase, since a TestCase wraps the test in a transaction, I would assume running `select_for_update` in the test and opening a new db connection would work fine, but I've found that in this scenario, the method under tests is unable to fetch _any_ from the database which I don't quite understand. In a TransactionTestCase this doesn't happen. Any ideas on what might be the problem?

millerdev replied:
Yes, the transaction management in `TestCase` is a little more complicated than that. It starts a transaction during `setUpClass`, then makes a SAVEPOINT before each test, and it rolls back to and releases the SAVEPOINT after each test. Finally, it rolls back the entire transaction in `tearDownClass` after all tests have completed.

`ROLLBACK TO SAVEPOINT` does not release locks acquired by `SELECT FOR UPDATE` because [that would violate the ACID properties of the transaction](https://www.postgresql.org/message-id/1154639749.24639.22.camel%40dogma.v10.wvs). Therefore, any effects of `SELECT FOR UPDATE` in one test will be visible to other tests in the same class that are run later.

> I think the `select_for_update` needs to be outside of the `new_db_connection`. The purpose of the `new_db_connection` is to ensure the method under test uses a different db connection from the one that holds the lock on a record

Ah yes, I get it. We could work around this with another context manager that restores the original connection once the `select_for_update` has run, but before closing the "new connection". This would be pretty easy to write, and it would work something like this:

```py
with new_db_connection() as cn:
    _ = UserReportingMetadataStaging.objects.select_for_update().get(pk=record.id)
    with cn.restore_previous_connection():
        _process_reporting_metadata_staging()
```

Let me know if you want help writing that.

Another way we could solve this problem without implementing `restore_previous_connection` would be to put this test in its own `TestCase` and document that it should be the only test in that class. That would be a bit hacky, but easy enough.

gherceg replied:
@millerdev and I paired on this and came up with this solution: https://github.com/dimagi/commcare-hq/pull/33745/commits/00ab97f2d0dd6bb7cef334fb3f582693891a481c. 

It isn't the prettiest, but allows us to continue to use a TestCase which is faster and has a safer cleanup mechanism than TransactionTestCase.

gherceg committed:
Test process_reporting_metadata_staging with locked record

This is an inelegant solution to testing behavior with a locked
record without needing to use a TransactionTestCase, which are
much slower than TestCase, and resort to truncating tables to
cleanup the database after the test run which can cause issues
with other tests. Think software plans and other sql objects
that are created as part of initial CommCare HQ setup.

gherceg committed:
Use patch directly instead of mock.patch

Solves conflict with es/sql-user-data

gherceg commented:
I set USER_REPORTING_METADATA_BATCH_ENABLED to true on staging and tested that this task still processes as expected with this change deployed, which it does. Merging.

gherceg commented:
I explain in the description of https://github.com/dimagi/commcare-hq/pull/33811 why the `ResourceConflict` sentry errors flagged above aren't a concern, but I still created a followup PR in an effort to further reduce the number of conflicts encountered in this task.
